pkg/network/server.go

```diff
@@ -331,12 +332,30 @@ func (s *Server) startProtocol(p Peer) {
 			err = p.WriteMsg(m)
 		case <-timer.C:
 			// Try to sync in headers and block with the peer if his block height is higher then ours.
-			if p.Version().StartHeight > s.chain.BlockHeight() {
+			if p.LastBlockIndex() > s.chain.BlockHeight() {
 				err = s.requestBlocks(p)
 			}
 			if err == nil {
 				timer.Reset(s.ProtoTickInterval)
 			}
+			if s.chain.HeaderHeight() >= p.LastBlockIndex() {
+				block, errGetBlock := s.chain.GetBlock(s.chain.CurrentBlockHash())
+				if errGetBlock != nil {
+					err = errGetBlock
+				} else {
+					diff := uint32(time.Now().UTC().Unix()) - block.Timestamp
+					if diff > uint32(s.PingInterval/time.Second) {
+						p.UpdatePingSent(true)
+						err = p.WriteMsg(NewMessage(s.Net, CMDPing, payload.NewPing(s.id, s.chain.HeaderHeight())))
+					}
+				}
+			}
+		case <-pingTimer.C:
+			if p.PingSent() && p.GetPingLimit() < 0 {
```

Comment:
Well, it makes this limit a lifetime limit, that is if there are more than three unanswered pings in the whole connection history we drop the connection. I don't think it's appropriate, you should rather count the number of outstanding pings and the number of successive failed pings, because there is also a possibility of the other side's late answers like when you've sent one ping that wasn't answered in time, you send another one and then get two replies, this situation should also be handled correctly.